### PR TITLE
Adding the Pixel model

### PR DESCRIPTION
**Pull request by ZihanWangKi**
# Description

> :memo: Please include a summary of the change. 
>  
> * Please also include relevant motivation and context.  
> * List any dependencies that are required for this change.  

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [x] A new research paper code implementation

## Tests

> :memo: Please describe the tests that you ran to verify your changes.
>  
> * Provide instructions so we can reproduce.  
> * Please also list any relevant details for your test configuration.  

**Test Configuration**:

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [x] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [x] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

File: official/projects/pixel/configs/pixel.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
+# Copyright 2022 The TensorFlow Authors. All Rights Reserved.
```
Comment (saberkun) [inline official/projects/pixel/configs/pixel.py:1]: 2023

File: official/projects/pixel/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# Language Modelling with Pixels (PIXEL)
+
+TF2 implementation of [PIXEL](https://arxiv.org/abs/2207.06991).
+
+### Setup
+The current setup requires a numpyfied pytorch pixel model and preprocessed data.
+We have converted such a pre-trained model (along with sst2 data) and uploaded it to
```
Comment (saberkun) [inline official/projects/pixel/README.md:7]: This gs is probably private. I think we should move them else where. Can you remove this for now and send us an email including the paths?

File: official/projects/pixel/modeling/pixel.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+from official.modeling import tf_utils
+from official.vision.modeling.backbones import vit
+from official.vision.modeling.layers import nn_blocks
+from official.vision.modeling.layers import nn_layers
+layers = tf.keras.layers
+
+class ViTEncoder(tf.keras.layers.Layer):
+  """
+  ViT Encoder, copied from official/vision/modeling/backbones/vit.Encoder.
+  Changed the inputs to accomodate masks.
```
Comment (saberkun) [inline official/projects/pixel/modeling/pixel.py:29]: Can you elaborate more? 
What are the masks?

File: official/projects/pixel/utils/convert_numpy_weights_to_tf.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import numpy as np
```
Comment (saberkun) [inline official/projects/pixel/utils/convert_numpy_weights_to_tf.py:1]: Please add the license of 2023 to this file.

File: official/projects/pixel/modeling/pixel.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+from official.modeling import activations
+from official.modeling import tf_utils
+from official.vision.modeling.backbones import vit
+from official.vision.modeling.layers import nn_blocks
+from official.vision.modeling.layers import nn_layers
+layers = tf.keras.layers
+
+class ViTEncoder(tf.keras.layers.Layer):
+  """
+  ViT Encoder, copied from official/vision/modeling/backbones/vit.Encoder.
```
Comment (CrickWu) [inline official/projects/pixel/modeling/pixel.py:28]: nit: Encoder in ...vit.py

File: official/projects/pixel/modeling/pixel.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        'attention_dropout_rate': self._attention_dropout_rate,
+        'kernel_regularizer': self._kernel_regularizer,
+        'inputs_positions': self._inputs_positions,
+        'init_stochastic_depth_rate': self._init_stochastic_depth_rate,
+        'kernel_initializer': self._kernel_initializer,
+        'add_pos_embed': self._add_pos_embed,
+        'pos_embed_origin_shape': self._pos_embed_origin_shape,
+        'pos_embed_target_shape': self._pos_embed_target_shape,
+    }
+    config.update(updates)
```
Comment (CrickWu) [inline official/projects/pixel/modeling/pixel.py:116]: Is the update necessary?

File: official/projects/pixel/tasks/classification.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+      return {
+          self.metric_type: sklearn_metrics.matthews_corrcoef(preds, labels)
+      }
+    elif self.metric_type == 'pearson_spearman_corr':
+      preds = np.reshape(preds, -1)
+      labels = np.reshape(labels, -1)
+      pearson_corr = stats.pearsonr(preds, labels)[0]
+      spearman_corr = stats.spearmanr(preds, labels)[0]
+      corr_metric = (pearson_corr + spearman_corr) / 2
+      return {self.metric_type: corr_metric}
```
Comment (CrickWu) [inline official/projects/pixel/tasks/classification.py:198]: nit: ending with a new line

File: official/projects/pixel/utils/convert_numpy_weights_to_tf.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  """Convert pixel model from huggingface to official.projects.pixel."""
+  num_layers = 12
+  num_attention_heads = 12
+  hidden_size = 768
+  patch_c = 3
+  patch_h = 16
+  patch_w = 16
+  head_size = hidden_size // num_attention_heads
+  assert head_size * num_attention_heads == hidden_size
+  # print(vit_encoder.encoder.patch_to_embed.weights)
```
Comment (CrickWu) [inline official/projects/pixel/utils/convert_numpy_weights_to_tf.py:22]: nit: remove?

File: official/projects/pixel/utils/convert_numpy_weights_to_tf.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+  model_name = json.load(open(f"{data_path}/model_name.json"))
+  model_params = np.load(open(f"{data_path}/model_param.npy", "rb"), allow_pickle=True)
+
+  config = vit_cls.ViTConfig()
+  task = vit_cls.ViTClassificationTask(config)
+  model = task.build_model()
+
+  convert(model, {k: v for k, v in zip(model_name, model_params)})
+  tf.train.Checkpoint(encoder=model.encoder).write(output_model_name)
```
Comment (CrickWu) [inline official/projects/pixel/utils/convert_numpy_weights_to_tf.py:87]: nit: ending with a new line

**Review by CrickWu — COMMENTED** [review]
LGTM.

**Review by saberkun — APPROVED** [review]
